### ISSUE TITLE
dynamo rewrite for allreduce, default group

### ISSUE THREAD

**Issue by wconstab** [pytorch/pytorch#112028]
### 🚀 The feature, motivation and pitch

motivated by this post
https://discuss.pytorch.org/t/functional-collectives-used-in-torch-compile/190627

TLDR is we currently have dynamo support to rewrite certain forms of non-functional allgather/reducescatter into functional ones, to make the program traceable.

we should probably be able to support the same rewrite for allreduce and possibly also infer the 'default' processgroup in case it is omitted from the arguments to the c10d collective.

cc @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @pragupta @msaroufim @dcci @aditvenk @xmfan @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @chenyang78 @aakhundov @voz

**Comment by jon-chuang** [comment]
Is this a duplicate of: https://github.com/pytorch/pytorch/issues/109774

**Comment by sssiva81** [comment]
@wconstab  Will you be adding support for rewrites for reduce collective calls also?  Looking to trace DeepSpeed which uses reduce collective in Stage2.

**Comment by wconstab** [comment]
@yf225 can you de-duplicate this with any other open issues and comment on the current plan?

**Comment by dolpm** [comment]
> [@dolpm](https://github.com/dolpm) after your torchcomms work, do we have a new general strategy- should we be able to trace all collectives (inplace, c10d) and their PG's / comms into the graph?

all torchcomms collectives should be traceable, but the work doesn't really have any implications on the tracability of any existing c10d ops. you can play around with the tests [here](https://github.com/meta-pytorch/torchcomms/blob/main/comms/torchcomms/tests/integration/py/FullgraphCompileTest.py) if you want to see fx graphs and stuff. 

being out-of-tree makes the dynamo integration a little hacky (lots of monkey patching), but i have plans to move it [into pytorch](https://github.com/pytorch/pytorch/pull/176463) which should clean it up a lot (i've just been busy with some other things these past few weeks).
